lib/rules/media-feature-name-unit-allowed-list/index.js

```diff
@@ -0,0 +1,99 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const validateOptions = require('../../utils/validateOptions');
+const { TokenType } = require('@csstools/css-tokenizer');
+const { isTokenNode } = require('@csstools/css-parser-algorithms');
+const {
+	isMediaFeaturePlain,
+	isMediaFeatureRange,
+	parse,
+} = require('@csstools/media-query-list-parser');
+const validateObjectWithArrayProps = require('../../utils/validateObjectWithArrayProps');
+const { isString } = require('../../utils/validateTypes');
+const matchesStringOrRegExp = require('../../utils/matchesStringOrRegExp');
+const flattenArray = require('../../utils/flattenArray');
+const atRuleParamIndex = require('../../utils/atRuleParamIndex');
+
+const ruleName = 'media-feature-name-unit-allowed-list';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (unit, name) => `Unexpected unit "${unit}" for name "${name}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/media-feature-name-unit-allowed-list',
+};
+
+/** @type {import('stylelint').Rule} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [validateObjectWithArrayProps(isString)],
+		});
+
+		if (!validOptions) {
+			return;
+		}
+
+		root.walkAtRules(/^media$/i, (atRule) => {
+			const mediaQueryList = parse(atRule.params);
+
+			mediaQueryList.forEach((mediaQuery) => {
+				mediaQuery.walk((entry) => {
+					if (!isMediaFeaturePlain(entry.node) && !isMediaFeatureRange(entry.node)) {
+						return;
+					}
+
+					const name = isMediaFeatureRange(entry.node)
+						? entry.node.name.name.value[1].toString()
+						: entry.node.name.toString();
+
+					const nameKey = Object.keys(primary).find((nameIdentifier) =>
+						matchesStringOrRegExp(name, nameIdentifier),
+					);
+
+					if (!nameKey) {
+						return;
+					}
+
+					const nameList = flattenArray(primary[nameKey]);
+
+					if (!nameList) {
+						return;
+					}
+
+					entry.node.walk((childEntry) => {
+						if (!isTokenNode(childEntry.node) || childEntry.node.value[0] !== TokenType.Dimension) {
+							return;
+						}
+
+						const unit = childEntry.node.value[4].unit;
```

Comment:
Another "hardcoded" value indicated by @romainmenke.

Reply:
Correct :)

The check on line 69 ensures that this is a dimension token.
A dimension token has extra data :

```js
{
  unit: "px",
  value: 10,
  type: "integer"
}
```

----

Index `4` is where all the usable and unescaped token values can be found.
This is the same for all token types.

```
[<token type>, <raw contents>, <start index>, <end index>, <parsed contents>]
```

Reply:
[suggest] I believe we can improve readability by using destructuring assignment like this:

```js
entry.node.walk(({ node: childNode }) => {
	if (!isTokenNode(childNode)) {
		return;
	}

	const [tokenType, , startIndex, endIndex, parsedContent] = childNode.value;

	if (tokenType !== TokenType.Dimension) {
		return;
	}
```